Add Segment Anything 2 (SAM2)

SangbumChoi opened the pull request:
# What does this PR do?

https://github.com/huggingface/transformers/issues/32308

As stated in this issue this PR is making SAM2 compatible to transformers

<!--
Congratulations! You've made it this far! You're not quite done yet though.

Once merged, your PR is going to appear in the release notes with the title you set, so make sure it's a great title that fully reflects the extent of your awesome contribution.

Then, please replace this with a description of the change and which issue is fixed (if applicable). Please also include relevant motivation and context. List any dependencies (if any) that are required for this change.

Once you're done, someone will review your PR shortly (see the section "Who can review?" below to tag some potential reviewers). They may suggest changes to make the code even better. If no one reviewed your PR after a week has passed, don't hesitate to post a new comment @-mentioning the same persons---sometimes notifications get lost.
-->

<!-- Remove if not applicable -->

Fixes # (issue)


## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [ ] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#create-a-pull-request),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?


## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

<!-- Your PR will be replied to more quickly if you can figure out the right person to tag with @

 If you know how to use git blame, that is the easiest way, otherwise, here is a rough guide of **who to tag**.
 Please tag fewer than 3 people.

Models:

- text models: @ArthurZucker
- vision models: @amyeroberts
- speech models: @sanchit-gandhi
- graph models: @clefourrier

Library:

- flax: @sanchit-gandhi
- generate: @zucchini-nlp (visual-language models) or @gante (all others)
- pipelines: @Narsil
- tensorflow: @gante and @Rocketknight1
- tokenizers: @ArthurZucker
- trainer: @muellerzr and @SunMarc

Integrations:

- deepspeed: HF Trainer/Accelerate: @muellerzr
- ray/raytune: @richardliaw, @amogkam
- Big Model Inference: @SunMarc
- quantization (bitsandbytes, autogpt): @SunMarc

Documentation: @stevhliu

HF projects:

- accelerate: [different repo](https://github.com/huggingface/accelerate)
- datasets: [different repo](https://github.com/huggingface/datasets)
- diffusers: [different repo](https://github.com/huggingface/diffusers)
- rust tokenizers: [different repo](https://github.com/huggingface/tokenizers)

Maintained examples (not research project or legacy):

- Flax: @sanchit-gandhi
- PyTorch: See Models above and tag the person corresponding to the modality of the example.
- TensorFlow: @Rocketknight1

 -->

SangbumChoi committed:
initial comment

SangbumChoi committed:
test

SangbumChoi committed:
Merge branch 'sam2' of https://github.com/SangbumChoi/transformers into sam2

SangbumChoi committed:
initial conversion for outline

SangbumChoi committed:
intermediate commit for configuration

RUFFY-369 committed:
chore:init files for sam2

SangbumChoi committed:
adding arbitary undefined config

SangbumChoi committed:
Merge pull request #1 from SangbumChoi/sam2_hf

SAM2 extra init changes

SangbumChoi committed:
check

SangbumChoi committed:
add vision

SangbumChoi committed:
make style

haithamkhedr committed:
init sam2 base model

haithamkhedr committed:
Fix imports

haithamkhedr committed:
Linting

RUFFY-369 committed:
chore:sam to sam2 classes

haithamkhedr committed:
Linting

haithamkhedr committed:
Add sam2 to models.__init__

RUFFY-369 committed:
chore:match prompt encoder with sam2 code

RUFFY-369 committed:
chore:prepare kwargs for mask decoder

SangbumChoi commented:
closed due to https://github.com/huggingface/transformers/pull/32394

SangbumChoi committed:
Merge pull request #3 from SangbumChoi/sam2_hf

Sam2 hf

SangbumChoi committed:
Merge pull request #2 from SangbumChoi/sam2_config

check

haithamkhedr committed:
Add image/video predictors

haithamkhedr committed:
Add CUDA kernel

haithamkhedr committed:
Add output classes

haithamkhedr committed:
linting

haithamkhedr committed:
Add logging info

RUFFY-369 commented:
@amyeroberts Can we continue this work on SAM2 as I think the authors' contributions have stopped. 
cc @SangbumChoi

SangbumChoi committed:
Merge branch 'main' of https://github.com/SangbumChoi/transformers into sam2

SangbumChoi committed:
Merge branch 'construct_sam2_base' of https://github.com/haithamkhedr/hf_transformers into sam2

SangbumChoi commented:
@RUFFY-369 I have reopened this PR let's get started from this branch since I have updated!

RUFFY-369 commented:
> @RUFFY-369 I have reopened this PR let's get started from this branch since I have updated!

Lets go :100:

amyeroberts commented:
@RUFFY-369 @SangbumChoi Excited to see this being picked up and the SAM-2 efforts revived! As there hasn't been any recent activity on #32394, we'll treat this as the active PR which will likely be merged in. 

Let us know if you need any help or have any questions getting this into the library!

giswqs commented:
Thank you very much for working on this! Can't wait to try this out

RUFFY-369 commented:
> @RUFFY-369 @SangbumChoi Excited to see this being picked up and the SAM-2 efforts revived! As there hasn't been any recent activity on #32394, we'll treat this as the active PR which will likely be merged in.
> 
> Let us know if you need any help or have any questions getting this into the library!

:100: :rocket:

SangbumChoi committed:
tmp commit

SangbumChoi committed:
docs for sam2

bhack commented:
Are we targeting 2 or 2.1?

SangbumChoi commented:
Afaik there is no architecture difference in 2.1 so we can aim both but 2.1 for first priority

SangbumChoi committed:
Merge branch 'main' of https://github.com/SangbumChoi/transformers into sam2

SangbumChoi committed:
enable image processing

SangbumChoi committed:
check difference of original SAM2
- difference is the order of ToTensor()
- please see https://pytorch.org/vision/main/_modules/torchvision/transforms/functional.html#resize

SangbumChoi committed:
enable promptencoder of sam2

SangbumChoi committed:
fix promprencoder

SangbumChoi committed:
Confirmed that PromptEncoder is exactly same (Be aware of bfloat16 and float32 difference)

SangbumChoi committed:
Confirmed that ImageEncoder is exactly same (Be aware the linting of init)

SangbumChoi committed:
Confirmed that MaskDecoder is exactly same (TO DO: lint variable name)

SangbumChoi committed:
SamModel is now available (Need more chore for name)

SangbumChoi committed:
make fix-copies

SangbumChoi committed:
make style

SangbumChoi committed:
make CI happy

SangbumChoi committed:
Merge branch 'main' of https://github.com/SangbumChoi/transformers into sam2

SangbumChoi committed:
Refactor VisionEncoder and PostioinEmbedding

SangbumChoi committed:
TO DO : fix the image_embeddings and sparse_embeddings part

SangbumChoi committed:
pure image inference done

SangbumChoi committed:
reusable features fix and make style

SangbumChoi committed:
Merge branch 'main' into sam2

SangbumChoi committed:
styling

SangbumChoi committed:
refactor memoryattention

SangbumChoi committed:
tmp

SangbumChoi committed:
tmp

SangbumChoi committed:
Merge branch 'main' of https://github.com/SangbumChoi/transformers into sam2

SangbumChoi committed:
refactor memoryencoder
TO DO : convert and inference the video pipeline

SangbumChoi committed:
TO DO : fix the image_encoder shape

SangbumChoi committed:
conversion finish
TO DO: need to check video inference

SangbumChoi commented:
@qubvel Hi Pavel, currently I'm finished with image + video checkpoint conversion and proceeding video inference pipeline.

For the image example I used to import with

`img_url = "https://huggingface.co/datasets/huggingface/documentation-images/resolve/main/transformers/model_doc/sam-car.png"
`

Is there any appropriate place to upload short video clip to use in docs and also conversion check?

qubvel commented:
Hi @SangbumChoi, a repo on the hub should be OK, alternatively, you can make a dataset. Then I can copy it to `hf-internal-testing` org. Let me know if you face any issues 🤗

neurohazardous commented:
Hello everyone! I wanted to know if this is an active PR and if there is any way I can help so that this gets deployed in transformers? Please let me know :)

SangbumChoi commented:
@neurohazardous Hi, It is an active PR, and already image based inference is done in this PR. What is left is video based inference. If you are familiar with transformers already then we can discuss about the collaboration!

SangbumChoi committed:
Merge branch 'main' into sam2

deepconvai commented:
Hi,

I really appreciate all the work within this PR and with all the other SAM2 attemps in other PRs, but I am wondering:
is it really a good goal to commit everything at once?
I mean there could be a phase 1, finishing, testing, merging SAM2's image based features, and continue in a new PR from there. 
It makes review easier, people get partial results sooner, etc., and when things needs to be rebased, it is less files, etc.
Given SAM2's improved speed, etc., an image-only model would be already a great addition. 
What do you think?

SangbumChoi commented:
@qubvel Do you have any additional comment for upper comment suggestion?

qubvel commented:
I think it might be a good option to deliver what we have as soon as possible. However, we must ensure that the API for the Image model will not be broken with the addition of the Video model, so it might be a bit tricky, as I'm not sure how tightly both of them are coupled.

SangbumChoi commented:
Yeah I also agree that it will be tricky and prefer to make both enabled.

hipsterusername commented:
Folks would greatly appreciate image support even if video isn’t there yet, for what it’s worth.

hlevring commented:
@SangbumChoi @qubvel @RUFFY-369  thanks for your work on this. Do you have some guesstimated time frame on when this PR may be usable for video predictions. 
 
 I was thinking to try to get something going with onnx runtime, but using transformers js with WebGPU would be much better ;)

qubvel commented:
Since there has been no recent activity in this PR, I suppose anyone from the community can reopen the PR and continue working on it. In case @SangbumChoi is fine with it, let's wait for his confirmation.

hipsterusername commented:
I’d propose that what is ready for image segmentation gets shipped if active work isn’t being done on video.

SangbumChoi committed:
Merge branch 'main' of https://github.com/SangbumChoi/transformers into sam2

SangbumChoi committed:
Merge branch 'main' into sam2

Commit:
make style

Commit:
remove video model

SangbumChoi committed:
lint

SangbumChoi committed:
Merge branch 'main' into sam2

SangbumChoi committed:
change

SangbumChoi committed:
python utils/check_docstringspy --check_all

SangbumChoi committed:
python utils/check_config_attributes.py

SangbumChoi committed:
remove copies for sam2promptencoder due to configuration

SangbumChoi commented:
@qubvel 

Hi Pavel, long time no see. Even though this PR is not perfectly ready. I have decided to make Image Part first.
There are only few things left.
- convert all model (only tiny atm)
- docstring update
- example code update

I know you are busy but you might want to roughly review this one :)

yonigozlan reviewed (commented):
Hi @SangbumChoi ! Thanks for the huge work on this.
I just made a quick pass for now and left a few comments mainly on things that need to be updated to follow new Transformers convention (init, no TensorFlow, attention implementation...). Since this PR was opened quite a long time ago, do you still have some time to work on it? Otherwise we can try to figure out how to take it from here.
Thanks again!

SangbumChoi commented:
> Hi @SangbumChoi ! Thanks for the huge work on this. I just made a quick pass for now and left a few comments mainly on things that need to be updated to follow new Transformers convention (init, no TensorFlow, attention implementation...). Since this PR was opened quite a long time ago, do you still have some time to work on it? Otherwise we can try to figure out how to take it from here. Thanks again!

@yonigozlan Hi, I will try!

SangbumChoi committed:
Merge branch 'main' into sam2

SangbumChoi committed:
change __init__.py

SangbumChoi committed:
remove tensorflow version

SangbumChoi committed:
fix that to not use direct comparison

SangbumChoi committed:
make style

SangbumChoi committed:
add missing import

SangbumChoi commented on src/transformers/models/sam2/modeling_sam2.py:
```diff
@@ -0,0 +1,2294 @@
+    return CUDA_KERNELS.get_connected_components(mask.to(torch.uint8).contiguous())
+
+
+def get_sdpa_settings():
+    if torch.cuda.is_available():
+        old_gpu = torch.cuda.get_device_properties(0).major < 7
+        # only use Flash Attention on Ampere (8.0) or newer GPUs
+        use_flash_attn = torch.cuda.get_device_properties(0).major >= 8
+        if not use_flash_attn:
+            warnings.warn(
+                "Flash Attention is disabled as it requires a GPU with Ampere (8.0) CUDA capability.",
+                category=UserWarning,
+                stacklevel=2,
+            )
+        # keep math kernel for PyTorch versions before 2.2 (Flash Attention v2 is only
+        # available on PyTorch 2.2+, while Flash Attention v1 cannot handle all cases)
+        pytorch_version = tuple(int(v) for v in torch.__version__.split(".")[:2])
+        if pytorch_version < (2, 2):
+            warnings.warn(
+                f"You are using PyTorch {torch.__version__} without Flash Attention v2 support. "
+                "Consider upgrading to PyTorch 2.2+ for Flash Attention v2 (which could be faster).",
+                category=UserWarning,
+                stacklevel=2,
+            )
+        math_kernel_on = pytorch_version < (2, 2) or not use_flash_attn
+    else:
+        old_gpu = True
+        use_flash_attn = False
+        math_kernel_on = True
+
+    return old_gpu, use_flash_attn, math_kernel_on
```
@yonigozlan 
Original version
```
        # Attention
        with torch.backends.cuda.sdp_kernel(
            enable_flash=USE_FLASH_ATTN,
            # if Flash attention kernel is off, then math kernel needs to be enabled
            enable_math=(OLD_GPU and dropout_p > 0.0) or MATH_KERNEL_ON,
            enable_mem_efficient=OLD_GPU,
        ):
            out = F.scaled_dot_product_attention(q, k, v, dropout_p=dropout_p)
```

Sam transformer version
```
        # Scaled dot product attention
        attn_mask = None
        if attention_similarity is not None:
            attn_mask = attention_similarity.unsqueeze(1).expand(-1, self.num_attention_heads, -1, -1)

        out = F.scaled_dot_product_attention(query, key, value, attn_mask=attn_mask)
```

enable_flash: Uses Flash Attention implementation (fastest, but requires compatible hardware)
enable_math: Uses standard mathematical implementation (most compatible, but slower)
enable_mem_efficient: Uses memory-efficient implementation (good balance of speed and memory usage)

```
import torch
import torch.nn.functional as F
import time

def test_sdp_kernel(enable_flash, enable_math, enable_mem_efficient, q, k, v):
    # Warm up
    for _ in range(5):
        with torch.backends.cuda.sdp_kernel(
            enable_flash=enable_flash,
            enable_math=enable_math,
            enable_mem_efficient=enable_mem_efficient
        ):
            _ = F.scaled_dot_product_attention(q, k, v)
    
    # Time measurement
    start_time = time.time()
    for _ in range(100):
        with torch.backends.cuda.sdp_kernel(
            enable_flash=enable_flash,
            enable_math=enable_math,
            enable_mem_efficient=enable_mem_efficient
        ):
            out = F.scaled_dot_product_attention(q, k, v)
    print(out)
    end_time = time.time()
    
    # Memory usage
    memory_allocated = torch.cuda.memory_allocated()
    memory_reserved = torch.cuda.memory_reserved()
    
    return {
        'time': (end_time - start_time) / 100,  # Average time per iteration
        'memory_allocated': memory_allocated,
        'memory_reserved': memory_reserved
    }

def run_comparison():
    # Test different combinations
    combinations = [
        {'name': 'Flash Only', 'flash': True, 'math': False, 'mem': False},
        {'name': 'Math Only', 'flash': False, 'math': True, 'mem': False},
        {'name': 'Mem Efficient Only', 'flash': False, 'math': False, 'mem': True},
        {'name': 'All Enabled', 'flash': True, 'math': True, 'mem': True},
        {'name': 'All Disabled', 'flash': False, 'math': False, 'mem': False}
    ]
    
    print("Testing different SDP kernel configurations...")
    print("=" * 80)
    
    # Create dummy tensors
    batch_size=2
    seq_len=2
    num_heads=1
    head_dim=4

    q = torch.randn(batch_size, num_heads, seq_len, head_dim, device='cuda')
    k = torch.randn(batch_size, num_heads, seq_len, head_dim, device='cuda')
    v = torch.randn(batch_size, num_heads, seq_len, head_dim, device='cuda')

    for combo in combinations:
        print(f"\nTesting {combo['name']}:")
        print(f"Flash: {combo['flash']}, Math: {combo['math']}, Mem Efficient: {combo['mem']}")
        
        try:
            results = test_sdp_kernel(
                enable_flash=combo['flash'],
                enable_math=combo['math'],
                enable_mem_efficient=combo['mem'],
                q=q,
                k=k,
                v=v
            )            
            print(f"Average time per iteration: {results['time']:.6f} seconds")
            print(f"Memory allocated: {results['memory_allocated'] / 1024**2:.2f} MB")
            print(f"Memory reserved: {results['memory_reserved'] / 1024**2:.2f} MB")
            
        except Exception as e:
            print(f"Error: {str(e)}")
        
        print("-" * 80)

if __name__ == "__main__":
    # Clear CUDA cache before testing
    torch.cuda.empty_cache()
    
    # Run the comparison
    run_comparison()
```

I think if we change the module into transformer version WDYT?

SangbumChoi committed:
Merge branch 'main' of https://github.com/SangbumChoi/transformers into sam2

SangbumChoi commented:
@qubvel @yonigozlan Hi can you review this PR except for the https://github.com/huggingface/transformers/pull/32317#discussion_r2010457443 ?

SangbumChoi committed:
Merge branch 'main' into sam2

SangbumChoi committed:
Merge branch 'main' into sam2

qubvel commented:
Hi @SangbumChoi, we will try to review this week!

yonigozlan commented:
Hi @SangbumChoi ! Sorry for the delay on this, I have started reviewing and will continue this week. I'm happy to help push updates to get up to speed with the current state of Transformers and to support the video pipeline if you don't have the bandwidth to iterate on this in the coming weeks. let me know!

SangbumChoi commented:
@yonigozlan I am little bit busy until June 5th but I can do my best after this. Current model is already converted the weight including memory encoder which is related to video pipeline. 

Let's do the video pipeline. :) I will summarize the TO DOs in this weekdays.

yonigozlan commented:
@SangbumChoi  Sounds good! I'll review + push some changes in the meantime then, let's see if I get to the video part by June 5 😅. Thanks again for your huge work, excited to merge this PR soon!

yonigozlan committed:
fix image_embedding_size

yonigozlan committed:
Merge remote-tracking branch 'upstream/main' into sam2

yonigozlan committed:
refactor Sam2 Attention

SangbumChoi committed:
Merge branch 'main' into sam2

yonigozlan committed:
add fully working video inference (refactoring todo)

yonigozlan committed:
Merge branch 'sam2' of https://github.com/SangbumChoi/transformers into sam2

yonigozlan committed:
clarify _prepare_memory_conditioned_features

yonigozlan committed:
simplify modeling code, remove unused paths

yonigozlan committed:
use one model

yonigozlan committed:
use auto_docstring

SangbumChoi commented:
@yonigozlan Hi, I think I can work on CVPR week (maybe coding while at the hotel :) ), I will start to understand the change that you have made. Is there anything specific detail that I should aware?

yonigozlan committed:
refactor rope embeddings

yonigozlan committed:
nit

yonigozlan commented:
Hey @SangbumChoi ! Main changes are the addition of the video pipeline. Changes from the original repo for now are mainly the attention implementation and RoPE Embeddings to be more inline with Transformers standards (still some refactoring to do for RoPE.
Still lots of refactoring left to do to see how to best handle the "inference state" as defined in the original video predictor implementation, to be somewhat in line with the usual image processor/video processor/model interaction in Transformers. I'll probably get to that early next week

yonigozlan committed:
not using multimask when several points given

SangbumChoi committed:
Merge branch 'main' of https://github.com/SangbumChoi/transformers into sam2

SangbumChoi committed:
add all sam2.1

SangbumChoi committed:
add video tmp

hipsterusername commented:
Will once again propose the split out of video and image as separate releases/PRs, given the state of this.

yonigozlan committed:
add Sam2VideoSessionState + fast image proc + video proc

yonigozlan committed:
Merge branch 'sam2' of https://github.com/SangbumChoi/transformers into sam2

yonigozlan committed:
remove init_states from model

yonigozlan committed:
fix batch inference

yonigozlan committed:
add image integration tests

yonigozlan committed:
Merge remote-tracking branch 'upstream/main' into sam2

yonigozlan committed:
Merge remote-tracking branch 'upstream/main' into sam2

yonigozlan committed:
uniformize modeling code with other sam models and use modular

yonigozlan committed:
pass vision tests an most model tests

yonigozlan committed:
All tests passing

yonigozlan committed:
add offloading inference state and video to cpu

yonigozlan committed:
fix inference from image embedding and existing mask

yonigozlan committed:
fix multi_boxes mask inference

yonigozlan committed:
Fix batch images + batch boxes inference

yonigozlan committed:
improve processing for image inference

yonigozlan committed:
add support for mask generation pipeline

yonigozlan committed:
add support for get_connected_components post processing in mask generation

yonigozlan committed:
add fast image processor sam, image processor tests and use modular for sam2 image processor

yonigozlan committed:
Merge remote-tracking branch 'upstream/main' into sam2

yonigozlan committed:
fix mistake in sam after #39120

yonigozlan committed:
fix init weights

SangbumChoi committed:
refactor convert

yonigozlan committed:
add integration tests for video + other improvements

yonigozlan committed:
add needed missing docstrings

SangbumChoi committed:
Merge pull request #4 from SangbumChoi/sam2_sbchoi

refactor convert

yonigozlan committed:
Improve docstrings and

yonigozlan committed:
Merge branch 'sam2' of https://github.com/SangbumChoi/transformers into sam2

yonigozlan committed:
improve inference speed by avoiding cuda sync

SangbumChoi committed:
Merge branch 'main' of https://github.com/SangbumChoi/transformers into sam2

SangbumChoi committed:
add test

SangbumChoi committed:
skip test for vision_model

SangbumChoi committed:
minor fix for vision_model

SangbumChoi committed:
fix vision_model by adding sam2model and change the torch dependencies

SangbumChoi committed:
remove patch_size

SangbumChoi committed:
remove image_embedding_size

SangbumChoi committed:
fix patch_size

SangbumChoi committed:
fix test

SangbumChoi committed:
make style

yonigozlan committed:
Separate hieradet and vision encoder in sam2

yonigozlan committed:
Merge branch 'sam2' of https://github.com/SangbumChoi/transformers into sam2

yonigozlan committed:
fixup

SangbumChoi commented:
@yonigozlan Hi Yoni! the CI seems passed. I will be working on docs and minor comments. Let me know if you need other thing to check.

yonigozlan commented:
> @yonigozlan Hi Yoni! the CI seems passed. I will be working on docs and minor comments. Let me know if you need other thing to check.

I think some comprehensive docs is the last thing missing! I'll open a separate PR for EdgeTAM as the differences are more important than I thought.
Thanks a lot for the fixes you pushed btw!

qubvel commented on src/transformers/models/sam2/modeling_sam2.py:
```diff
@@ -0,0 +1,3633 @@
+        """
+        self.images = list(video)
+        self.num_frames = len(video)
+        self.inference_device = inference_device
+        self.video_storage_device = video_storage_device
+        self.inference_state_device = inference_state_device
+        self.async_loading_frames = async_loading_frames
+        self.video_height = video_height
+        self.video_width = video_width
+        self.device = video.device
```
I would expect a `State` object to store `data`, not params. These are more related to input args or config?
Alternatively, it can be named as a `Cache` to store cache params + it should use it while updating cache internal **state**, e.g. when you put some objects to cache you move them to a proper device.

Not a req, just a design thought, so happy to discuss, WDYT?

yonigozlan replied:
Indeed `Cache` might be a more appropriate name, although video_height video_width and the different devices might be weird to have in a cache?
Also if I understand correctly, you mean the "moving to device" logic should be internal to the Cache/State object itself?

yonigozlan replied:
Here what would be the other parameters? I'll remove async_loading_frames, and the ones I see left would be video_height/width, torch_dtype and maybe inference_device?
Agreed that they are not really cache params, but not sure if it's worth having another object passed around between processing and modeling code just for these?

SangbumChoi replied:
Or following variable is just one time declaration for the given video tensor so we can also group into some kind of 
```
metadata: dict = {
  torch_dtype, inference_device
  num_frames: int = None
  video_height: int = None
  video_width: int = None
}
```

qubvel commented on tests/models/sam2/test_modeling_sam2.py:
```diff
@@ -0,0 +1,1422 @@
+        ann_frame_idx = 0  # the frame index we interact with
+        ann_obj_id = 1  # give a unique id to each object we interact with (it can be any integers)
+
+        inference_state = self.processor.process_new_points_or_box_for_video_frame(
+            inference_state=inference_state,
+            frame_idx=ann_frame_idx,
+            obj_ids=ann_obj_id,
+            input_points=[[[[210, 350], [250, 220]]]],
+            input_labels=[[[1, 1]]],
+        )
```
type annotations for input_points and input_labels are not correct, as well as return typehint

qubvel commented on tests/models/sam2/test_modeling_sam2.py:
```diff
@@ -0,0 +1,1422 @@
+
+        # test propagate in video frames
+        frames = []
+        for frame_idx, out_mask_logits in self.model.propagate_in_video(
+            inference_state=inference_state,
+            start_frame_idx=ann_frame_idx,
+            max_frame_num_to_track=2,
+        ):
+            frames.append(out_mask_logits)
+        frames = torch.stack(frames, dim=0)
```
how can we add a point on intermediate frame?

yonigozlan replied:
I think we can do this we minimal changes! And you're right it would be great to have.
I'll try to add something, but I imagine it more like this:

```python
video_inference_cache = None
frame_idx = 0
for frame in stream:
    if frame_idx = 0:
        boxes = detector(frame)
        boxes_with_ids = tracker(boxes)
        processed_frame = processor(frame)
        video_inference_cache = processor.process_new_points_or_box_for_video_frame(boxes_with_ids, frame_idx=0, obj_ids=boxes_with_ids.ids)
        sam2_output = sam2_model.infer_on_video_frame_with_new_inputs(video_inference_cache, frame=processed_frame, frame_idx=0, obj_ids=boxes_with_ids.ids, stream=True)
    else:
        processed_frame = processor(frame)
        sam2_output = sam2_model.propagate_in_video(video_inference_cache, frame=processed_frame , stream=True)
```

As passing in inputs (boxes or points) at each step kind of defeats the purpose of sam2.
WDYT?

SangbumChoi replied:
BTW if this is the proposed method then it would be great to have high priority because this might affect to the docs or inference jupyter.

yonigozlan replied:
Yes although it only affects inference on streamed videos, the rest of the docs shouldn't be affected

qubvel reviewed (commented):
HUGE work done, kudos @SangbumChoi and @yonigozlan 🚀 🚀 🚀  
I completed an initial review, mostly style and code quality comments. I have also raised some design questions, which you are welcome to discuss either below or on Slack 🤗

yonigozlan commented on src/transformers/models/sam2/configuration_sam2.py:
```diff
@@ -0,0 +1,704 @@
+        self.non_overlap_masks = non_overlap_masks  # whether to apply non-overlapping constraints on output masks
+
+
+__all__ = [
+    "Sam2Config",
+    "Sam2HieraDetConfig",
+    "Sam2VisionConfig",
+    "Sam2PromptEncoderConfig",
+    "Sam2MaskDecoderConfig",
+    "Sam2MemoryAttentionConfig",
+    "Sam2MemoryEncoderConfig",
+]
```
Yes 😅, some don't correspond to a module that can be instantiated by itself (Sam2PromptEncoderConfig, Sam2MaskDecoderConfig etc.) so wasn't sure if we should make them into separate configs, but that's how it was done for SAM

qubvel replied:
Yeah, it might be worth merging some of them in case it does not lead to code bloating

yonigozlan commented on src/transformers/models/sam2/image_processing_sam2_fast.py:
```diff
@@ -0,0 +1,847 @@
+                    mask_flat = mask.flatten(0, 1, 2).unsqueeze(1)
+                else:
+                    raise ValueError("Input masks should be a list of `torch.tensors` or a list of `np.ndarray`")
+                if torch.cuda.is_available():
+                    try:
+                        load_cuda_kernels()
+                    except Exception as e:
+                        print(f"Could not load custom CUDA kernels for postprocessing: {e}")
```
Not sure I understand, it should be cached and not loaded every time as this is in the init?

SangbumChoi replied:
@yonigozlan I think 
> it should be cached and not loaded every time as this is in the init?

This is the correct comprehension

yonigozlan commented on src/transformers/models/sam2/modeling_sam2.py:
```diff
@@ -0,0 +1,3633 @@
+            raise NotImplementedError(f"Unsupported data format: {self.data_format}")
+        self.normalized_shape = (normalized_shape,)
+
+    def forward(self, x: torch.Tensor) -> torch.Tensor:
+        if self.data_format == "channels_last":
+            x = torch.nn.functional.layer_norm(x, self.normalized_shape, self.weight, self.bias, self.eps)
+        elif self.data_format == "channels_first":
+            input_dtype = x.dtype
+            x = x.float()
+            u = x.mean(1, keepdim=True)
+            s = (x - u).pow(2).mean(1, keepdim=True)
+            x = (x - u) / torch.sqrt(s + self.eps)
+            x = x.to(dtype=input_dtype)
+            x = self.weight[:, None, None] * x + self.bias[:, None, None]
+        return x
```
This is from modular sam and copied itself from convnext. I can make the changes here but maybe out of scope for this PR?

yonigozlan commented on src/transformers/models/sam2/modeling_sam2.py:
```diff
@@ -0,0 +1,3633 @@
+            torch_dtype (`torch.dtype`, *optional*, defaults to `torch.float32`):
+                The torch dtype to use for the video.
+        """
+        self.images = list(video)
```
Don't remember why I did that but I think we should be able to you're right!

yonigozlan committed:
review changes part 1

yonigozlan committed:
remove MemoryEncoderConfig and MemoryAttentionConfig

yonigozlan committed:
pass q_stride instead of q_pool module

SangbumChoi commented on src/transformers/models/sam2/modeling_sam2.py:
```diff
@@ -0,0 +1,3633 @@
+        # For video sequence inference
+        self.memory_attention = Sam2MemoryAttention(config.memory_attention_config)
+        self.memory_encoder = Sam2MemoryEncoder(config.memory_encoder_config)
+
```
Actually original SAM2.1 also load Memory related weight even though it is the image based inference. However, to be honest could be option.
```
class SamImageModel():
        self.shared_image_embedding = Sam2PositionalEmbedding(config.prompt_encoder_config)
        # For single image inference
        self.vision_encoder = AutoModel.from_config(config.vision_config)
        self.prompt_encoder = Sam2PromptEncoder(config.prompt_encoder_config)
        self.mask_decoder = Sam2MaskDecoder(config.mask_decoder_config)
class SamVideoMdoel(SamImageModel):
        # For video sequence inference
        self.memory_attention = Sam2MemoryAttention(config.memory_attention_config)
        self.memory_encoder = Sam2MemoryEncoder(config.memory_encoder_config)
```
User have to load SamImageModel if they want to load Image only inference. But I also agree that this might cause some confusion.

yonigozlan committed:
add inference on streamed videos

yonigozlan committed:
explicitely process streamed frames

yonigozlan committed:
nit

yonigozlan commented:
@qubvel @SangbumChoi Just pushed changes to add inference support on streamed videos, you can look at the new test here to see how it would work:
https://github.com/huggingface/transformers/blob/f75e04d5e43e37f7f3a5d02450c7abb04b4ee90b/tests/models/sam2/test_modeling_sam2.py#L1368-L1421

yonigozlan committed:
Improve docstrings in Sam2Model

qubvel commented:
Hey @yonigozlan, here is an example of API. I think it should be much closer to our standards and simpler for users, let me know if you think it's doable

```py

video_model = Sam2VideoModel.from_pretrained("...")

video_processor = AutoVideoProcessor.from_pretrained("...")
inference_state = video_processor.init_video_session(inference_device=torch_device) 

raw_video = prepare_video() 

for frame_idx, frame in enumerate(raw_video): 
    
    inputs = video_processor(images=frame, device=torch_device, return_tensors="pt") 
    
    if frame_idx == 0: 
        inference_state = video_processor.update_inference_state( 
            inference_state, 
            frame_idx=frame_idx,
            # maybe pass them as dict objects? not super clear because of lots of nested lists
            obj_ids=1, 
            input_points=[[[[210, 350], [250, 220]]]], 
            input_labels=[[[1, 1]]],
            
            # do we need to pass it for each frame?
            original_size=inputs.original_sizes[0], 
        )

    with torch.no_grad():
        outputs = video_model(**inputs, inference_state=inference_state)
    


    video_res_masks.append(outputs.segment) 

```

qubvel commented on src/transformers/models/sam2/image_processing_sam2_fast.py:
```diff
@@ -0,0 +1,847 @@
+                    except Exception as e:
+                        print(f"Could not load custom CUDA kernels for postprocessing: {e}")
+                try:
+                    if max_hole_area > 0:
+                        mask = _fill_holes(mask_flat, mask, max_hole_area, mask_threshold)
+                    if max_sprinkle_area > 0:
+                        mask = _fill_sprinkles(mask_flat, mask, max_sprinkle_area, mask_threshold)
+                    processed_masks.append(mask)
```
ok, I guess we can use it from cv2/skimage, but it can be added in a follow-up PR if requested

yonigozlan committed:
update sam2 modeling with better gestion of inference state and cache, and separate Sam2Model and Sam2VideoModel

yonigozlan committed:
improve video inference api

yonigozlan committed:
change inference_state to inference_session

yonigozlan commented:
Hello @qubvel and @SangbumChoi ! Made a lot of changes to the PR, mainly on the video inference API, following @qubvel advice.
It should be much more transformers-like and intuitive now :)!
@SangbumChoi that does mean that the video examples in the docs will need to be changed, sorry about that. Once again you can have a look at all the video integration tests to see the new API.

yonigozlan committed:
Merge remote-tracking branch 'upstream/main' into sam2

yonigozlan commented:
run_slow: sam2

yonigozlan committed:
use modular for Sam2Model

qubvel reviewed (commented):
Just a few additional comments, some might be not relevant after our discussion 😄 
I will add more later!

yonigozlan committed:
fix convert sam2 hf

SangbumChoi committed:
Merge branch 'main' of https://github.com/SangbumChoi/transformers into sam2

SangbumChoi committed:
modular

SangbumChoi committed:
Update src/transformers/models/sam2/video_processing_sam2.py

Co-authored-by: Pavel Iakubovskii <qubvel@gmail.com>

SangbumChoi committed:
fix minor config

SangbumChoi committed:
Merge branch 'sam2' of https://github.com/SangbumChoi/transformers into sam2

SangbumChoi committed:
fix attention loading error

yonigozlan committed:
update modeling tests to use hub checkpoints

yonigozlan committed:
Use CI A10 runner for integration tests values + higher tolerance for video integration tests

yonigozlan commented on src/transformers/models/sam2/processing_sam2.py:
```diff
@@ -0,0 +1,785 @@
+        self,
+        images: ImageInput = None,
+        segmentation_maps: ImageInput = None,
+        input_points: Optional[
+            Union[list[float], list[list[float]], list[list[list[float]]], list[list[list[list[float]]]], torch.Tensor]
+        ] = None,
+        input_labels: Optional[Union[int, list[int], list[list[int]], list[list[list[int]]], torch.Tensor]] = None,
+        input_boxes: Optional[Union[list[float], list[list[float]], list[list[list[float]]], torch.Tensor]] = None,
+        original_sizes: Optional[Union[list[list[float]], torch.Tensor]] = None,
```
Agreed that the typing looks weird, but each of these inputs make sense in some use case imo.
i'm not sure it would be the best user experience to force inputs to be something like [[[[x, y]]]] for one input point on one image when we can easily infer what the user wants to do with [x,y]?

yonigozlan committed:
PR review part 1

SangbumChoi commented:
@yonigozlan Hi I modified some changes due to attention code style has been changed.

https://github.com/huggingface/transformers/commit/a5e2429ac464233b6acc08ee4104f12783f33d0f

yonigozlan commented:
@SangbumChoi Saw that thanks for fixing after merge!

yonigozlan committed:
fix doc

yonigozlan committed:
nit improvements

yonigozlan committed:
enforce one input format for points, labels and boxes

yonigozlan committed:
nit

yonigozlan committed:
last few nits from PR review

yonigozlan committed:
fix style

SangbumChoi committed:
fix the input type

SangbumChoi committed:
fix docs

SangbumChoi committed:
add sam2 model as conversion script

SangbumChoi commented:
Hi @yonigozlan @qubvel @ArthurZucker @Cyrilvallez. I just updated docs and the simple dtype update, and convert script. I think except for the Pavel's comment that on github (I think Yoni is keep dealing with it) 99% is done. Let me know if team need anything else. It would be great if we can merge this.

I also attached additional link that team have to look
### weight files
- https://huggingface.co/danelcsb/sam2.1_hiera_tiny (descriptions are written in here only)
- https://huggingface.co/danelcsb/sam2.1_hiera_small
- https://huggingface.co/danelcsb/sam2.1_hiera_base_plus
- https://huggingface.co/danelcsb/sam2.1_hiera_large
- https://huggingface.co/danelcsb/sam2_hiera_tiny
- https://huggingface.co/danelcsb/sam2_hiera_small
- https://huggingface.co/danelcsb/sam2_hiera_base_plus
- https://huggingface.co/danelcsb/sam2_hiera_large

Note that torch.allclose is not defined yet for the SAM2 (SAM2.1 is finished)

### notebook

[This](https://github.com/user-attachments/files/21452025/huggingface_video_predictor_example.1.ipynb.zip) notebook should be linked to the [here](https://github.com/huggingface/transformers/pull/32317/files#r2233740356).

yonigozlan committed:
improve sam2 doc

yonigozlan committed:
Merge branch 'sam2' of https://github.com/SangbumChoi/transformers into sam2

yonigozlan reviewed (approved):
Hey @SangbumChoi ! Thanks for updating the doc, I also added some more usage examples to reflect what we have in tests and notebooks

yonigozlan committed:
Merge remote-tracking branch 'upstream/main' into sam2

yonigozlan committed:
nit fixes + optimization

ArthurZucker reviewed (commented):
HUGE PR, we need to split the modalities, and this will make it easier to refactor / rethink the caching!

yonigozlan committed:
split sam2 and sam2_video in two models

yonigozlan commented on src/transformers/models/sam2/modular_sam2.py:
```diff
@@ -0,0 +1,4153 @@
+        # e.g. if fpn_top_down_levels is [2, 3], then only outputs of level 2 and 3
+        # have top-down propagation, while outputs of level 0 and level 1 have only
+        # lateral features from the same backbone level.
+        if config.fpn_top_down_levels is None:
+            # default is to have top-down features on all levels
```
No indeed!

yonigozlan commented on src/transformers/models/sam2/modular_sam2.py:
```diff
@@ -0,0 +1,4153 @@
+    def __init__(
+        self,
+        config: Sam2HieraDetConfig,
+        dim: int,
+        dim_out: int,
+        num_attention_heads: int,
+        query_stride: Optional[tuple[int, int]] = None,
```
Agreed in general but that would mean lists of 48 elements in the config for each parameter...

yonigozlan commented on src/transformers/models/sam2/modular_sam2.py:
```diff
@@ -0,0 +1,4153 @@
+    def __init__(
+        self,
+        config: Sam2HieraDetConfig,
+        dim: int,
+        dim_out: int,
+        num_attention_heads: int,
+        mlp_ratio: float = 4.0,
+        drop_path: float = 0.0,
+        query_stride: Optional[tuple[int, int]] = None,
+        window_size: int = 0,
+    ):
```
Same issue with the 48 elements needed in the config

yonigozlan committed:
PR review part 1

SangbumChoi committed:
Merge branch 'main' into sam2

SangbumChoi committed:
fix None for default slow processor of sam2

yonigozlan committed:
remove unecessary code path in sam2_video

yonigozlan committed:
Merge branch 'sam2' of https://github.com/SangbumChoi/transformers into sam2

yonigozlan committed:
refactor/simplify RoPE

yonigozlan committed:
replace embedding module list with embedding matrix

yonigozlan committed:
fix tests

yonigozlan committed:
remove kernel

yonigozlan commented on utils/check_repo.py:
```diff
@@ -191,6 +193,7 @@
     "models/bark/test_modeling_bark.py",
     "models/shieldgemma2/test_modeling_shieldgemma2.py",
     "models/llama4/test_modeling_llama4.py",
+    "models/sam2_video/test_modeling_sam2_video.py",
```
Added sam2_video here fow now as it doesn't follow typical forward format + is mostly tested in sam2

yonigozlan committed:
nit

yonigozlan committed:
Merge remote-tracking branch 'upstream/main' into sam2

SangbumChoi committed:
Merge branch 'main' into sam2

yonigozlan committed:
use lru_cache for sine_pos_embeddings

yonigozlan committed:
Merge branch 'sam2' of https://github.com/SangbumChoi/transformers into sam2

yonigozlan committed:
reorder sam2_video methods

yonigozlan committed:
simplify sam2_video

ArthurZucker reviewed (commented):
Thanks! There are still some pain points but great work!

yonigozlan commented on src/transformers/models/mask2former/modeling_mask2former.py:
```diff
@@ -855,18 +856,25 @@ def __init__(
+    @compile_compatible_method_lru_cache(maxsize=10)
+    def forward(
+        self,
+        shape: torch.Size,
+        device: Union[torch.device, str],
+        dtype: torch.dtype,
+        mask: Optional[Tensor] = None,
```
Opened a separate PR here, I'll add a 🚨

yonigozlan commented on src/transformers/models/sam2_video/modular_sam2_video.py:
```diff
@@ -0,0 +1,2812 @@
+        return self.pos_embeddings
+
+
+def rotate_half(x):
```
Not exactly the same I think, more here: https://github.com/huggingface/transformers/pull/32317#discussion_r2258399344

yonigozlan commented on src/transformers/models/sam2_video/modular_sam2_video.py:
```diff
@@ -0,0 +1,2812 @@
+        cos, sin = self.rotary_emb()
+
+        # Apply rotary position encoding, excluding some keys if specified
+        if num_k_exclude_rope > 0:
```
Yes

yonigozlan commented on src/transformers/models/sam2_video/modular_sam2_video.py:
```diff
@@ -0,0 +1,2812 @@
+            self.temporal_positional_encoding_projection_layer = torch.nn.Identity()
+
+        self.occlusion_spatial_embedding_parameter = None  # compatibility with Sam2
+        if config.enable_occlusion_spatial_embedding:
```
This is True for Sam2.1 and False for Sam2 checkpoints

SangbumChoi commented on src/transformers/models/sam2/modular_sam2.py:
```diff
@@ -0,0 +1,1469 @@
+
+if is_torch_available():
+    import torch
+    from torch.nn import functional as F_t
```
Doesnt just F is the standard way?

SangbumChoi commented on src/transformers/models/sam2/modular_sam2.py:
```diff
@@ -0,0 +1,4153 @@
+            self.num_mask_tokens,
+            config.iou_head_depth,
+            activation=config.feed_forward_hidden_act,
+            sigmoid_output=True,
```
Yeah, other method is to divide the Sam2FeedForward with respecting function, but maybe just more nit with config + additional parameter should be the minimum change that we can make.

yonigozlan committed:
PR review part 1

yonigozlan committed:
simplify sam2 video a lot

yonigozlan committed:
more simplification

yonigozlan committed:
update integration tests with updated conftest

yonigozlan committed:
more explicit config for hieradet

yonigozlan committed:
do post_processing outside of sam2 video model

yonigozlan committed:
Improve Sam2VideoVisionRotaryEmbedding

yonigozlan committed:
fix tests

yonigozlan committed:
Merge remote-tracking branch 'upstream/main' into sam2

yonigozlan committed:
update docs and fix mask2former/oneformer

yonigozlan committed:
avoid unnecessary reshapes/permute

yonigozlan commented:
Hello @SangbumChoi @qubvel and @ArthurZucker !
I refactored sam2 video almost completely, and sam2 to a lesser degree. Now sam2 video is muuuch simpler and more transformers like (post processing is now done outside of the model when it's possible), while keeping the same functionalities as the original implementation. Hopefully it is now also easier to review and maintain!
I would love another round of reviews (hopefully the last one), and it would be great if we can get this merged by the end of the week!

yonigozlan committed:
fix device concatenating points

SangbumChoi reviewed (commented):
For me I understood the major change. I will comment more when the arthur's comment is finshed and making some examples 👍🏼

ArthurZucker reviewed (approved):
thanks for pushing the limits! With such a big modeling code, making it readable is very hard, but very important! Kudos to both of you 🤗

yonigozlan committed:
small dtype fix

yonigozlan commented on src/transformers/models/sam2/image_processing_sam2_fast.py:
```diff
@@ -739,52 +664,39 @@ def post_process_masks(
-            else:
-                processed_masks = masks
-            masks = processed_masks
+        # if max_hole_area > 0 or max_sprinkle_area > 0:
+        #     processed_masks = []
+        #     TODO: add connected components kernel for postprocessing
+        #     for mask in masks:
+        #         if mask.ndim == 3:
+        #             mask_flat = mask.flatten(0).unsqueeze(1)
+        #         elif mask.ndim == 4:
+        #             mask_flat = mask.flatten(0, 1).unsqueeze(1)
+        #         elif mask.ndim == 5:
+        #             mask_flat = mask.flatten(0, 1, 2).unsqueeze(1)
+        #         else:
+        #             raise ValueError("Input masks should be a list of `torch.tensors` or a list of `np.ndarray`")
+        #     else:
+        #         processed_masks = masks
+        #     masks = processed_masks
```
Yes indeed, I'll mark that as a todo and remove it here. This uses the connected_component cuda kernels, but I haven't had time to look into adding it to the kernels library. However it seems to have very minimal impact, so I don't think it's a must-have before merging

yonigozlan commented on src/transformers/models/sam2_video/modular_sam2_video.py:
```diff
@@ -0,0 +1,2375 @@
+        self.v_proj = nn.Linear(self.kv_in_dim, self.internal_dim)
+        self.o_proj = nn.Linear(self.internal_dim, self.hidden_size)
+
+        self.rotary_emb = Sam2VideoVisionRotaryEmbedding(config=config)
```
Since I need to get the rotary embeddings in the forward of this module, how would I get them if they are in a parent module?

yonigozlan committed:
PR review

yonigozlan committed:
nit

yonigozlan committed:
Merge remote-tracking branch 'upstream/main' into sam2

yonigozlan committed:
fix style and finish up doc

yonigozlan committed:
fix style

yonigozlan committed:
fix docstrings

yonigozlan committed:
Merge remote-tracking branch 'upstream/main' into sam2

yonigozlan committed:
fix modular

yonigozlan commented:
Hi @SangbumChoi yes than would be great! I'll update the model doc and checkpoints in the docs as well once I have an answer to see if we can have the checkpoints in the facebook org on the hub. I can add a link to your notebooks then!

BodhiHu commented:
Hello,
Great works here :D
Do you guys know which release will include SAM2 ?

Thank you

BodhiHu commented:
Thanks a lot 😄
